src/extra/components/adq.py

```diff
@@ -0,0 +1,1294 @@
+            interleaving samples or not, only needed if it could not be
+            detected automatically.
+        clock_ratio (int, optional): Digitizer sampling clock as
+            multiple of the bunch repetition rate (4.5 MHz).
+        pulses (extra.components.pulses.PulsePattern): Pulse component
+            to pull pulse information, by default a
+            [XrayPulses][extra.components.XrayPulses] object is
+            constructed unless pulse information are explicitly disabled
+            by passing `False`. Most pulse separation operations require
+            this to be present.
+        first_pulse_offset (int, optional): Sample where the first
+            pulse begins, 10000 by default.
+        single_pulse_length (int, optional): Samples per pulse for
+            the case all trains only contain a single pulse.
+        preprocess_kw (dict, optional): Custom settings for raw data
+            preprocessing done by this component, please see arguments
+            of [AdqChannel.preprocess].
+    """
+
+    # 4.5 Mhz, see extra.components.pulses.
+    _bunch_repetition_rate = 1.3e9 / 288
+
+    _adq_pipeline_re = re.compile(r'(\w+)\/ADC\/(\d):network')
+    _adq_channel_re = re.compile(
+        r'digitizers.channel_(\d)_([A-Z]).raw.samples')
+
+    # Set of digitizers using 3G boards with a 392 clock ratio that do
+    # not advertise as such.
+    _3g_digitizer = {'SQS_DIGITIZER_UTC2/ADC/1'}
+
+    def __init__(self, data, channel, digitizer=None, pulses=None,
+                 first_pulse_offset=10000, single_pulse_length=25000,
+                 interleaved=None, clock_ratio=None, preprocess_kw={}):
+        if digitizer is None or digitizer not in data.instrument_sources:
+            digitizer = self._find_adq_pipeline(data, digitizer or '')
+
+        self._instrument_src = data[digitizer]
+
+        if len(channel) == 2:
+            self._channel_board = int(channel[0])
+            self._channel_letter = channel[1].upper()
+        elif len(channel) == 3:
+            self._channel_board = int(channel[0])
+            self._channel_letter = channel[2].upper()
+        else:
+            raise ValueError('channel expected to be 2 or 3 characters, '
+                             'e.g. 1A or 1_A')
+
+        self._channel_number = ord(self._channel_letter) - ord('A')
+        self._channel_name = f'{self._channel_board}_{self._channel_letter}'
+
+        key = f'digitizers.channel_{self._channel_name}.raw.samples'
+
+        if key not in self._instrument_src:
+            raise ValueError(f'key {key} for channel {channel} not found '
+                             f'in digitizer instrument source {digitizer}')
+
+        self._raw_key = self._instrument_src[key]
+
+        # Try to find control source.
+        device_id = digitizer[:digitizer.index(':')]
+
+        if device_id in data.control_sources:
+            self._control_src = data[device_id]
+        else:
+            self._control_src = None
```

Comment:
This component will work with only the `INSTRUMENT` data, while the enclosing `AdqDigitizer` will not. There are currently DAQ problems preventing universal acquisition of slow data for ADQ digitizers, but CTRL is working on that.